src/wallet/walletdb.cpp

```diff
@@ -298,6 +301,40 @@ bool WalletBatch::EraseLockedUTXO(const COutPoint& output)
     return EraseIC(std::make_pair(DBKeys::LOCKED_UTXO, std::make_pair(output.hash, output.n)));
 }
 
+bool WalletBatch::WriteHDKey(const CExtKey& extkey)
+{
+    std::vector<unsigned char> xpub(BIP32_EXTKEY_SIZE);
+    extkey.Neuter().Encode(xpub.data());
+
+    CPrivKey privkey = extkey.key.GetPrivKey();
+
+    return WriteIC(std::make_pair(DBKeys::HDKEY, xpub), std::make_pair(privkey, PrivKeyChecksum(privkey, xpub)), false);
+}
+
+bool WalletBatch::WriteHDCryptedKey(const CExtPubKey& extpub, const std::vector<unsigned char>& crypted_key)
+{
+    // Compute a checksum of the encrypted key
+    uint256 checksum = Hash(crypted_key);
+
+    std::vector<unsigned char> xpub(BIP32_EXTKEY_SIZE);
+    extpub.Encode(xpub.data());
+
+    const auto key = std::make_pair(DBKeys::HDCKEY, xpub);
+    if (!WriteIC(key, std::make_pair(crypted_key, checksum), false)) {
+        return false;
+    }
+    EraseIC(std::make_pair(DBKeys::HDKEY, xpub));
```

Comment:
37527c9c6930dd86d5d012e94d2e08d6a207d240 could add a comment here that we're deleting the unencrypted key. Or make the intention more readable:

```cpp
const auto db_key_plain_text = std::make_pair(DBKeys::HDKEY, xpub);
const auto db_key_encrypted = std::make_pair(DBKeys::HDCKEY, xpub);
if (!WriteIC(db_key_plain_text, std::make_pair(crypted_key, checksum), false)) {
    return false;
}
EraseIC(db_key_unencrypted);

```

I find it a worrying that no test caught the fact that we weren't deleting this in an earlier version of the PR (b88337daba). Can we introduce a test that looks for unencrypted key material e.g. by just checking _every_ record?

Reply:
Added a test that should catch that in the future.